File: android/controller/src/main/java/org/openbot/controller/DriveCommandReducer.kt
package org.openbot.controller

import android.util.Log
import org.openbot.controller.customComponents.DualDriveSeekBar
import kotlin.math.absoluteValue

object DriveCommandReducer {
    private var lastRight = 0f
    private var lastLeft = 0f
    private const val withinRange = .02f

    fun filter(rightValue: Float, leftValue: Float) {
        if (isDifferent(rightValue, leftValue)) {
            lastLeft = leftValue
            lastRight = rightValue
            val msg = "{driveCmd: {r:${rightValue}, l:${leftValue}}}"
            ConnectionSelector.getConnection().sendMessage(msg)
        }
    }

    private fun isDifferent (right: Float, left: Float): Boolean {
        if ((left - lastLeft).absoluteValue <= withinRange && (right - lastRight).absoluteValue <= withinRange) {
            return false
        }
        return true
    }
}

File: controller/flutter/lib/screens/driveCommandReducer.dart
import 'package:openbot_controller/globals.dart';
import 'package:openbot_controller/utils/forwardSpeed.dart';

class DriveCommandReducer {
  static double lastRight = 0;
  static double lastLeft = 0;
  static double withinRange = .02;

  static void filter(double rightValue, double leftValue) {
    if (isDifferent(rightValue, leftValue)) {
      lastLeft = leftValue;
      lastRight = rightValue;
      String msg =
          "{driveCmd: {r:${rightValue.toPrecision(2)}, l:${leftValue.toPrecision(2)}}}";
      clientSocket?.writeln(msg);
    }
  }

  static bool isDifferent(double right, double left) {
    if ((left - lastLeft).abs() <= withinRange &&
        (right - lastRight).abs() <= withinRange) {
      return false;
    }
    return true;
  }
}
